File: sdk-js/src/transport.ts
// Copyright 2026 Harsha Krishne Gowda
// SPDX-License-Identifier: Apache-2.0

/** TokenBudget event transport — batched delivery using fetch(). */

import type { UsageEvent, ResolvedConfig } from './types.js';

/**
 * Event transport that batches events and flushes them periodically.
 * Uses the built-in fetch() API (Node 18+ / browsers).
 * Never blocks the main thread — fire and forget.
 */
export class EventTransport {
  private queue: UsageEvent[] = [];
  private config: ResolvedConfig;
  private timer: ReturnType<typeof setInterval> | null = null;
  private stopped = false;

  constructor(config: ResolvedConfig) {
    this.config = config;
    this.startFlushTimer();
  }

  private startFlushTimer(): void {
    if (this.timer) return;
    this.timer = setInterval(() => {
      this.flush().catch(() => {
        // silently ignore flush errors
      });
    }, this.config.flushIntervalMs);

    // Allow Node.js to exit even if the timer is still running
    if (typeof this.timer === 'object' && 'unref' in this.timer) {
      this.timer.unref();
    }
  }

  /**
   * Enqueue an event. Never blocks, never throws.
   */
  send(event: UsageEvent): void {
    if (this.stopped) return;
    if (this.queue.length >= this.config.maxQueueSize) {
      // Drop the event — queue is full
      return;
    }
    this.queue.push(event);

    // Auto-flush if batch size reached
    if (this.queue.length >= this.config.batchSize) {
      this.flush().catch(() => {
        // silently ignore flush errors
      });
    }
  }

  /**
   * Flush all queued events to the API endpoint.
   */
  async flush(): Promise<void> {
    if (this.queue.length === 0) return;

    // Drain the queue into a local array
    const events = this.queue.splice(0);
    const url = `${this.config.endpoint}/v1/events/batch`;
    const payload = { events };

    try {
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          Authorization: `Bearer ${this.config.apiKey}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(payload),
      });

      if (!response.ok) {
        // Re-queue events for retry
        this.requeue(events);
      }
    } catch {
      // Re-queue events for retry on next flush
      this.requeue(events);
    }
  }

  private requeue(events: UsageEvent[]): void {
    for (const event of events) {
      if (this.queue.length >= this.config.maxQueueSize) break;
      this.queue.push(event);
    }
  }

  /**
   * Stop the flush timer, flush remaining events, and clean up.
   */
  async shutdown(): Promise<void> {
    this.stopped = true;
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
    await this.flush();
  }

  /** Visible for testing. */
  get queueLength(): number {
    return this.queue.length;
  }
}

File: sdk/tokenbudget/transport.py
# Copyright 2026 Harsha Krishne Gowda
# SPDX-License-Identifier: Apache-2.0

"""TokenBudget event transport — async batched delivery to the TokenBudget API."""
from __future__ import annotations

import dataclasses
import logging
import queue
import threading
from typing import TYPE_CHECKING

import httpx

if TYPE_CHECKING:
    from tokenbudget.config import TokenBudgetConfig
    from tokenbudget.types import UsageEvent

logger = logging.getLogger(__name__)


def _event_to_dict(event: "UsageEvent") -> dict:
    """Serialize a UsageEvent to a JSON-serializable dict."""
    return dataclasses.asdict(event)


class EventTransport:
    """Thread-safe event transport with background flushing."""

    def __init__(self, config: "TokenBudgetConfig") -> None:
        self.config = config
        self.queue: queue.Queue["UsageEvent"] = queue.Queue(maxsize=config.max_queue_size)
        self._client = httpx.Client(
            headers={
                "Authorization": f"Bearer {config.api_key}",
                "Content-Type": "application/json",
            },
            timeout=10.0,
        )
        self._lock = threading.Lock()
        self._flush_thread: threading.Thread | None = None
        self._stop_event = threading.Event()
        self._start_flush_thread()

    def _start_flush_thread(self) -> None:
        """Start the background flush thread if not already running."""
        with self._lock:
            if self._flush_thread is None or not self._flush_thread.is_alive():
                self._flush_thread = threading.Thread(
                    target=self._flush_loop,
                    daemon=True,
                    name="tokenbudget-flush",
                )
                self._flush_thread.start()

    def _flush_loop(self) -> None:
        """Background loop: flush every flush_interval seconds."""
        while not self._stop_event.wait(timeout=self.config.flush_interval):
            self._do_flush()

    def send(self, event: "UsageEvent") -> None:
        """Enqueue an event. Never blocks, never raises."""
        try:
            self.queue.put_nowait(event)
        except queue.Full:
            logger.warning("TokenBudget event queue is full; dropping event.")
        except Exception as exc:
            logger.warning("TokenBudget: failed to enqueue event: %s", exc)

    def flush_sync(self) -> None:
        """Drain the queue and POST all events to the API endpoint."""
        self._do_flush()

    def _do_flush(self) -> None:
        """Internal flush implementation."""
        if self.queue.empty():
            return

        # Drain the queue into a local list
        events: list["UsageEvent"] = []
        while True:
            try:
                events.append(self.queue.get_nowait())
            except queue.Empty:
                break

        if not events:
            return

        url = f"{self.config.endpoint}/v1/events/batch"
        payload = {"events": [_event_to_dict(e) for e in events]}

        try:
            response = self._client.post(url, json=payload)
            response.raise_for_status()
        except Exception as exc:
            logger.warning("TokenBudget: flush failed (%s); re-queuing %d events.", exc, len(events))
            # Re-queue events for retry on next flush
            for event in events:
                try:
                    self.queue.put_nowait(event)
                except queue.Full:
                    logger.warning("TokenBudget: queue full during re-queue; dropping event.")

    def shutdown(self) -> None:
        """Stop the flush thread, flush remaining events, and close the HTTP client."""
        self._stop_event.set()
        if self._flush_thread and self._flush_thread.is_alive():
            self._flush_thread.join(timeout=5.0)
        self._do_flush()
        try:
            self._client.close()
        except Exception:
            pass
